Add KBI0001 on shared git repositories

mslw opened the pull request:
This adds a note that I wrote after one of the previous office hours, with information on how owner checks for the top-level directory introduced with Git v2.35.2 lead to error messages when cloning other people's datasets, and why setting the safe.directory config option is required to make it work (but also should be a conscious decision).

Essentially migrating this: https://github.com/psychoinformatics-de/datalad-officehours/issues/8 in to here.

Wasn't sure what the discussion link should be - this PR? or specifically created issue in this repo?

mih commented:
Thanks for writing that down!

> Wasn't sure what the discussion link should be - this PR? or specifically created issue in this repo?

Yes, the PR seems optimal.

mih commented:
I kept thinking about a desirable scope for this KBI. In the process I found https://github.com/datalad/datalad/issues/7270, which is clearly related. However, here the "sharedness" is someone different (it is merely a readable directory, with possibly no desire/intent for sharing by the owner). So maybe the scope of "accessing shared datasets", or maybe "permission and security considerations for sharing datasets".

mslw commented:
> I kept thinking about a desirable scope for this KBI. In the process I found [datalad/datalad#7270](https://github.com/datalad/datalad/issues/7270), which is clearly related. However, here the "sharedness" is someone different (it is merely a readable directory, with possibly no desire/intent for sharing by the owner). So maybe the scope of "accessing shared datasets", or maybe "permission and security considerations for sharing datasets".

Agree - the intention was to explain the "Git refuses to operate in this repository" error (i.e. "git directory outside of the directories owned by the current user") - I will try to come up with a better title.

bpoldrack commented:
Just one thing from my POV:

I believe it would be good to include the git message `Git refuses to operate in this repository, probably because it is owned by someone else.` literally in the document, so that this can be found by searching for it.

mslw commented:
Repeating comment from inline suggestion by @mih 

> it is unclear to me, whether this needs further clarification that it is not about clones from bare repos.

I think we need to investigate a little bit. At first glance with git version 2.36.0:

```
❱ datalad clone "ria+file:///data/group/psyinf/dataset_store#946e8cac-432b-11ea-aac8-f0d5bf7b5561" throwaway
[INFO   ] initremote() takes 1 positional argument but 2 were given                                                                           
install(error): /data/group/psyinf/sfb1451/throwaway (dataset) [CommandError: 'git -c diff.ignoreSubmodules=none for-each-ref '--format=%(refname:strip=2)' refs/heads refs/remotes' failed with exitcode 128 [err: 'fatal: unsafe repository ('/data/group/psyinf/dataset_store/946/e8cac-432b-11ea-aac8-f0d5bf7b5561' is owned by someone else)
To add an exception for this directory, call:

	git config --global --add safe.directory /data/group/psyinf/dataset_store/946/e8cac-432b-11ea-aac8-f0d5bf7b5561']]
```

But it might as well be more about the parent directory ownership, and less about this being bare or not...

mslw commented:
Okay, I tried to make a quick wrap, and a writer's block forced me had to re-think. This is my conclusion for now. Mostly a re-phrasing of what was said above. Feel free to correct me if I'm in a bad mindset.

- The scope is: operations that trigger Git's "unsafe repository" mechanism, i.e. DataLad operations that involve Git working in a directory owned by someone else. So, multi-user machines. That absolutely includes clone that's apparently read-only.
- Setting the `safe.directory` option must be left on the table for the users. We should suggest caution, but can't suggest users to abandon such use cases outright (even the [Handbook](https://handbook.datalad.org/en/latest/basics/101-116-sharelocal.html) covers cloning between users).
- Conclusion wording is a bit unfortunate and needs a tweak, but in the end the KBI does provide (cite) enough information about the logic behind Git's changes to be meaningful.
- If bare repos are unaffected, we should copy [this comment](https://github.com/psychoinformatics-de/knowledge-base/pull/1#discussion_r1162510688) advising againist using shared repositories almost verbatim (maybe expanding on what it means to have "own clones" - that's a mental shortcut for cloning from central location and not each other)
- But if bare repos are also affected by a mere virtue of sitting in a shared directory (like `/data`) - which I now think is the case but need to verify - then the admonition about avoiding shared setups is out of scope - a bare repo will also need to be trusted.
- Both title and conclusion need to be updated depending on the point above.

Therefore, the only **blocker** for me is whether a clone from a bare repo would be affected. For this I have to check again, and I ran out of time to do it now.

I will go back to it, but if you (unspecific "you, the reader") read this and can run a quick test, please let me know.

mslw commented:
Yup, bare repo doesn't matter. `/data/group/psyinf/inputstore` is owned by another user (we are both part of the same group).

```
❱ git --version                                                                             
git version 2.30.2
❱ datalad clone "ria+file:///data/group/psyinf/inputstore#47a9ef9e-9fa0-49cd-873f-c51cc501da96" oldgit
[INFO   ] scanning for unlocked files (this may take some time)                                                                               
install(ok): /home/mszczepanik/Documents/test-clone/oldgit (dataset)
mszczepanik@juseless in ~/Documents/test-clone
```

```
❱ git --version                                                                                       
git version 2.36.0
❱ datalad clone "ria+file:///data/group/psyinf/inputstore#47a9ef9e-9fa0-49cd-873f-c51cc501da96" newgit
install(error): /home/mszczepanik/Documents/test-clone/newgit (dataset) [CommandError: 'git -c diff.ignoreSubmodules=none for-each-ref '--format=%(refname:strip=2)' refs/heads refs/remotes' failed with exitcode 128 [err: 'fatal: unsafe repository ('/data/group/psyinf/inputstore/47a/9ef9e-9fa0-49cd-873f-c51cc501da96' is owned by someone else)
To add an exception for this directory, call:

	git config --global --add safe.directory /data/group/psyinf/inputstore/47a/9ef9e-9fa0-49cd-873f-c51cc501da96']] [CommandError: 'git -c diff.ignoreSubmodules=none for-each-ref '--format=%(refname:strip=2)' refs/heads refs/remotes' failed with exitcode 128 [err: 'fatal: unsafe repository ('/data/group/psyinf/inputstore/47a/9ef9e-9fa0-49cd-873f-c51cc501da96' is owned by someone else)
To add an exception for this directory, call:

	git config --global --add safe.directory /data/group/psyinf/inputstore/47a/9ef9e-9fa0-49cd-873f-c51cc501da96']]
```

mslw commented on kbi/0001/index.rst:
As I realised in the discussion below, although the commit quoted from the Git project refers mostly to shared repos, it is only the clone location that matters for the (un)safe directory. So a bare repo (shared or non-shared) in any kind of shared ("scratch") space would trigger the warning. So the option must be set, or clone won't happen.

I rephrased the recommendation to emphasize that it is a declaration of trust.

I also added a single commit rephrasing your comment about shared repositories above - although in (new) context this may seem slightly tangential.

mslw commented:
I rephrased the recommendation, changed the title, and added a short comment about shared mode (if deemed out of scope, the comment is a single commit that can be reverted). From my point of view this KBI is ready to merge, but given the changes I'll leave it until at least one review.

mih commented on kbi/0001/index.rst:
```diff
@@ -0,0 +1,187 @@
+.. index::
+   single: datalad; clone
+   single: config; safe.directory
```
```suggestion
.. index:: datalad; clone, git config; safe.directory
```

Mainly adding the `git` to `config`.

mih commented on kbi/0001/index.rst:
```diff
@@ -0,0 +1,187 @@
+
+This technical note provides background information about the observed
+behavior, and guidance on possible solutions.
+
```
```suggestion


Background
----------

```

What comes next is already background, and I think there is no need for "further background" below.

mih commented on kbi/0001/index.rst:
```diff
@@ -0,0 +1,187 @@
+Related GitHub blog post:
+https://github.blog/2022-04-12-git-security-vulnerability-announced/
+
+Further background
+------------------
+
```
```suggestion
```

See comment above.

mih commented on kbi/0001/index.rst:
```suggestion
Since ``safe.directory`` is a Git security mechanism, DataLad will not
set the option automatically, and the decision is left to the user.
```

mih reviewed (approved):
I left a few, I believe, no-brainer comments.
